torch/_dynamo/variables/distributed.py

```diff
@@ -112,10 +117,20 @@ def call_method(
     ) -> "VariableTracker":
         from . import ConstantVariable
 
-        allowed_methods = ["__init__", "__setattr__"]
-        # placement types dynamo tracking allows only __init__
-        # and __setattr__ methods, the latter is for case like `Shard(dim)`
-        if name in allowed_methods:
+        # Placement types dynamo tracking only allows following methods
+        # and __setattr__  is for case like `Shard(dim)` and methods.
+        # Methods in the list must satisfy:
+        #    1. Input arguments are constants and do not need to be guarded on;
+        #    2. Output is constant with respect to their inputs
+        allowed_constant_methods = [
+            "__init__",
```

Comment:
Can you add one example of tracing ```__init__```, I think we should review this carefully since it's probably triggers side effect. In your unit tests, the ```Shard(0)``` and  ```Replicate``` are generated outside of the compile region, I think it won't trigger the tracing of ```__init__```.